### PR TITLE
Introduce option allowing TLS 1.3 server to prefer psk_ke over psk_dhe_ke

### PR DESCRIPTION
**Pull request by minichma**
Introduce option `SSL_OP_PREFER_NO_DHE_KEX`, which allows configuring a TLS1.3 server to prefer session resumption using PSK-only key exchange over PSK with DHE, if both are available. The option is ignored unless PSK-only is explicitly allowed via `SSL_OP_ALLOW_NO_DHE_KEX`.

* Config option name: `PreferNoDHEKEX`
* Server flag: `prefer_no_dhe_kex`

This PR addresses #22783.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

##### ToDo
- [x] add version of first release to the HISTORY section of `SSL_CONF_cmd.pod`.

### CODE REVIEW

File: ssl/statem/extensions_srvr.c [hunk 1]
```diff
@@ -1645,12 +1645,22 @@ EXT_RETURN tls_construct_stoc_key_share(SSL_CONNECTION *s, WPACKET *pkt,
...
         return EXT_RETURN_NOT_SENT;
     }
-    if (s->hit && (s->ext.psk_kex_mode & TLSEXT_KEX_MODE_FLAG_KE_DHE) == 0) {
-        /*
-         * PSK ('hit') and explicitly not doing DHE (if the client sent the
-         * DHE option we always take it); don't send key share.
+    if (s->hit) {
+        /* PSK ('hit') */
+
+        /* If we're doing PSK ('hit') but the client doesn't support psk-dhe,
```
Comment (t8m) [inline ssl/statem/extensions_srvr.c:1651]: Nit: Multiline comments need to start with `/*` and end with `*/` on separate lines.

File: ssl/statem/extensions_srvr.c [hunk 2]
```diff
@@ -1645,12 +1645,22 @@ EXT_RETURN tls_construct_stoc_key_share(SSL_CONNECTION *s, WPACKET *pkt,
...
+         * we don't need to send a key share.
          */
-        return EXT_RETURN_NOT_SENT;
+        if ((s->ext.psk_kex_mode & TLSEXT_KEX_MODE_FLAG_KE_DHE) == 0)
+            return EXT_RETURN_NOT_SENT;
+
+        /* If both, psk_ke and psk_dh_ke are available, we do psk_dh_ke and
```
Comment (t8m) [inline ssl/statem/extensions_srvr.c:1657]: As above

**Review by t8m — CHANGES_REQUESTED** [review]
Good work. Could you please add CHANGES.md entry?

**Comment by minichma** [comment]
@t8m Thank you! Improved as suggested. As the doc build complained about docs missing in `openssl-s_(server|client).pod` I added the options there as well (and in `perlvars.pm` too). However, as the option is ignored on the client (which _could_ change in the future though), I'm not sure it is a good idea to add it to the client docs. What do you think? How would I remove it from the client docs without the linter complaining?

Btw, the CLA is on its way.

**Comment by t8m** [comment]
I would keep it at the client docs. However you also need to add the description for the new command line options in SSL_CONF_cmd.pod around line 98.

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by minichma** [comment]
@t8m I closed and reopened the PR to get rid of the 'CLA missing' tag. I've added the changelog in anticipation of version 3.3. Anything else missing from my side?

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
> @t8m I closed and reopened the PR to get rid of the 'CLA missing' tag. I've added the changelog in anticipation of version 3.3. Anything else missing from my side?

Nope, this is good as is.

**Comment by mattcaswell** [comment]
Pushed. Thanks!

**Pull request closed without merging**